pkg/node/node_address.go

```diff
@@ -402,3 +415,82 @@ func GetNodeAddressing(enableIPv4 bool) *models.NodeAddressing {
 		},
 	}
 }
+
+func getCiliumHostIPsFromFile(nodeConfig string) (ipv4GW, ipv6Router net.IP) {
+	f, err := os.Open(nodeConfig)
+	switch {
+	case err != nil:
+	default:
+		defer f.Close()
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			txt := scanner.Text()
+			switch {
+			case strings.Contains(txt, "IPV4_GATEWAY"):
+				// #define IPV4_GATEWAY 0xee1c000a
+				defineLine := strings.Split(txt, " ")
+				if len(defineLine) != 3 {
+					continue
+				}
+				ipv4GWHex := strings.TrimPrefix(defineLine[2], "0x")
+				ipv4GWint64, err := strconv.ParseInt(ipv4GWHex, 16, 0)
+				if err != nil {
+					continue
+				}
+				bs := make([]byte, net.IPv4len)
+				byteorder.NetworkToHostPut(bs, uint32(ipv4GWint64))
+				ipv4GW = net.IPv4(bs[0], bs[1], bs[2], bs[3])
+			case strings.Contains(txt, " ROUTER_IP "):
+				// #define ROUTER_IP 0xf0, 0xd, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, 0x8a, 0xd6
+				defineLine := strings.Split(txt, " ROUTER_IP ")
+				if len(defineLine) != 2 {
+					continue
+				}
+				ipv6 := common.C2GoArray(defineLine[1])
+				if len(ipv6) != net.IPv6len {
+					continue
+				}
+				ipv6Router = net.IP(ipv6)
+			}
+		}
+	}
+	return
+}
+
+// getCiliumHostIPsFromNetDev returns the first IPv4 link local and returns
+// it
+func getCiliumHostIPsFromNetDev(devName string) (ipv4GW, ipv6Router net.IP) {
+	hostDev, err := netlink.LinkByName(devName)
+	if err != nil {
+		return
+	}
+	addrs, err := netlink.AddrList(hostDev, netlink.FAMILY_ALL)
+	if err != nil {
+		return
+	}
+	for _, addr := range addrs {
+		if addr.IP.To4() != nil {
+			if addr.Scope == int(netlink.SCOPE_LINK) {
+				ipv4GW = addr.IP
+			}
+		} else {
+			if addr.Scope != int(netlink.SCOPE_LINK) {
+				ipv6Router = addr.IP
+			}
+		}
+	}
+	return
+}
+
+// getCiliumHostIPs returns the Cilium IPv4 gateway and router IPv6 address from
+// the node_config.h file if is present; or by deriving it from cilium_host
+// interface, on which only the IPv4 is possible to derive.
+func getCiliumHostIPs() (ipv4GW, ipv6Router net.IP) {
+	nodeConfig := option.Config.GetNodeConfigPath()
+	ipv4GW, ipv6Router = getCiliumHostIPsFromFile(nodeConfig)
+	if ipv4GW != nil {
+		return
+	}
+	ipv4GW, ipv6Router = getCiliumHostIPsFromNetDev(HostDevice)
+	return
```

Comment:
We don't do these naked returns in our code too often. It might still be good to return the variables directly despite it being redundant (I do prefer named return variables, however!)